### ISSUE TITLE
Allow structurally valid locators to be valid. 

### ISSUE THREAD

**Issue by johnjbarton** [SeleniumHQ/selenium#3056]
## Meta -
OS:  All

Selenium Version:  2.53.2

Browser:  All

For nodejs version.

In by.js we check for a valid locator:

https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/lib/by.js#L258

The check fails if the locator instance argument is created from a copy of the By class located in a different instance of the selenium-webdriver module.

Given nodejs node_modules layout, if anyone writes code that wraps selenium-webdriver, two instance of selenium-webdriver is relatively easy to occur  For example, protractor setups can easily result in two modules.

Since the vast majority of the selenium-webdriver feature set consists of function calls, tolerating two module instance -- however distasteful aesthetically -- is not difficult practically. 

The only exception I've found so far is in by.js, where an instanceof check is used. The instanceof is not essential to the logic, and a structural check would also work, eg

```js
function check(locator) {
  if (locator instanceof By || typeof locator === 'function') {
    return locator;
  }
  // Allow structurally equivalent locators, in case the argument
  // comes from an instance of By in another node module.
  if (typeof locator.using === 'string' && typeof locator.value === 'string') {
    return locator;
  }
  for (let key in locator) {
    if (locator.hasOwnProperty(key) && By.hasOwnProperty(key)) {
      return By[key](locator[key]);
    }
  }
  throw new TypeError('Invalid locator in by.js');
}
```

**Comment by jleyba** [comment]
This breaks encapsulation and weds us to the use of the using and value properties (which really shouldn't be publicly accessible).

The instanceof check should only fail if there are competing versions of selenium-webdriver in use, which underscores the issues with lack of encapsulation.

Finally, object literals are already supported if there's a key that maps to a locator strategy.  For example, `{id: 'foo'}` is equivalent to `By.id('foo')`

**Comment by jleyba** [comment]
Another instance where using conflicting versions will create problems for you is `selenium-webdriver`'s use of symbols to define internal methods used to serialize objects:

https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/chrome.js#L655

**Comment by johnjbarton** [comment]
Sorry, my description of this issue was incomplete. Let me try again to explain what I think is an issue.

Webdriver tests need to create a connection to the browser-driver, eg createSesssion(), then call some functions to operate on the remote browser to complete the test. As tests or applications become more complex, it's common practice to build up some additional 'helper' functions.  Consider these two lines from [a webdriver tutorial.](https://code.tutsplus.com/tutorials/an-introduction-to-webdriver-using-the-javascript-bindings--cms-21855)
```js
var webdriver = require('selenium-webdriver');
var browser = new webdriver.Builder().usingServer().withCapabilities({'browserName': 'chrome' }).build();
``` 
Now imagine a more complicated set of capabilities and set up of other kinds.  A development team may well decide to follow solid software engineering practice combine all of the session set up in a nodejs module. If that team also applies the common, standard, and recommended approach to nodejs development, they will make selenium-webdriver a dependency of their new setup-module. This combination of normal and common procedures will result in the failures such "Invalid Locator". 
The setup-module will have one copy of selenium-webdriver in its node_modules directory and the tests will have another.

The cause of this error will be quite mysterious and require someone with nodejs experience to debug.  Of course you could say "oh you should not do that", but I guess developers will. I did and a lot of other folks did too. 

Note that the potential hazard here does not uniformly impact the API. So perhaps we could have a conflict due to symbols used to serialize, but this is may, in practice, be very *unlikely*: developers are likely to call into the serialization code with only one module, the one used to create the webdriver session. But  developers are *likely* to hit the invalid locator because they obtain the locator from a different module. Thus fixing the issue for type-testing arguments may solve the vast majority of these cases.

On the plus side the system does not seem to use module state, only module-based types.  So solving this issue may not be too hard.  Webdriver has a reputation for being mysterious and solving this issue is a way to improve the status.

A reasonable alternative could be to throw if a module detects that is instantiated twice. Of course this will break some code and some of that code will be legit.
